### ISSUE TITLE
RFC: Monad transformers

### ISSUE THREAD

**Issue by ollef** [llvm-hs/llvm-irbuilder#4]
Currently we have a concrete `IRBuilder s` monad, where the `s` is a phantom type signifying what level we're at (Top level, function level, or block level).

I'd like to generalise this to a monad transformer so that users can bring their own transformer stacks. Ideally, I think I'd like to use an `mtl`-style `MonadIRBuilder` class to minimise lifting. The problem is that I haven't figured out how to do the strafification to different levels in that style.

Any ideas?

**Comment by ollef** [comment]
Although I'm reluctant to do it because I think it's quite nice to have my current plan is to remove the stratification such that it's easier to later turn into a `MonadIRBuilder` class, which means that:

* The `block` function gets type `block :: Name -> IRBuilder Name`, i.e. it's at the same level as instructions and also doesn't take a block builder argument. If you want indentation for blocks, you can do e.g.

```haskell
    blk1 <- block "b1"; do
      a <- fadd c1 c1
      b <- fadd a a
      c <- add c2 c2
      br blk2
```

As a sidenote, this way of doing blocks will make adapting the `sixten` backend easier as it uses basically this design.

* Calls to `function` can be nested when doing generation, with the semantics that if you define `g` while defining `f`, `g` comes before `f` in the resulting generated list of definitions.

* It's now legal (type-wise) to emit instructions at the top-level. These could either be ignored or yield a runtime error.

**Comment by sdiehl** [comment]
There's basically two alternative ways of implementing this. One would be using an indexed monad with that enforced the stratification at the monad operator level (i.e. ``(>>=) :: Monad m => m s a -> (a -> m s b) -> m s b`` ) but requires `-XRebindableSyntax` which is a bit much. Or due what the Kaleidoscope tutorial does and have a seperator transformer for each level of stratification and have the `toplevel, method, block` unpack the level into the outer level state. 

I quite like the phantom monad approach currently implemented because it doesn't really require anything more clever than us just controlling the exports of the module to make the interface type-safe. We use this same approach in [`llvm-hs-typed`](https://github.com/llvm-hs/llvm-hs-typed). 

As it's currently implemented it's just an mtl `MonadState` that can live happily at the bottom of another more complex stack.

**Comment by sdiehl** [comment]
If you fancy making a new branch with a new approach for the monad design, go for it. :+1:

**Comment by ollef** [comment]
Here's what it looks like without the stratification: https://github.com/llvm-hs/llvm-irbuilder/tree/RemoveTypeStratification

But I think I have a new idea that'll work more generally and which includes stratification where it matters.

First of all, we merge the function level and the block level. This can be done safely by assuming that we are in e.g. the `"entry"` block when a function starts.

We can either use the type `block :: Name -> IRBuilder Name` or the old `block :: Name -> IRBuilder r -> IRBuilder Name`. With the former `block` acts like an instruction. With the latter we'd allow nested blocks that we flatten in the generated code.

In addition to simplifying the implementation of this library it seems to me that merging the function and the block level means easier code-generation (think e.g. about how you'd compile an expression AST containing an `IfThenElse Expr Expr Expr` node; with the stratified approach you'd be forced to make the compilation function a `Function` level function and then you'd have to generate at least one block per AST node that you compile (and a few more when compiling `IfThenElse`), whereas you can get away with far fewer without stratification).

Next, we make a transformer and associated class for the function+block level, let's say `IRBuilderT` and `MonadIRBuilder`. We also add run functions to `IRBuilderT` that produce LLVM `Definition`s.

On top of that we have an _optional_ top level, let's call it `ToplevelIRBuilderT`, that has the `function` function takes an `IRBuilderT` as an argument. Here's where the stratification _actually_ matters and where we keep it; to disallow emitting instructions at the top level.

In summary this means that the function-level `MonadIRBuilder` can be used on its own and more flexibly than the current stratified approach. In particular, it can have IO or ST at the bottom of the stack which I want for Sixten, and it doesn't have to be at the top of the stack either.

For those that want the top level and can live with the less flexible monad stack that it entails there is the `TopLevelIRBuilder`. It'll also be used to write cute examples. :)

**Comment by sdiehl** [comment]
As far as I can tell this *removes* the type safety enforced by the extra parameter that enforces the emitter context?

**Comment by ollef** [comment]
It removes the block/function distinction (but I argue it's not necessary) but keeps the function/toplevel distinction (which _is_ necessary to ensure we can't emit instructions outside function bodies).

**Comment by ollef** [comment]
I've implemented that now if you want to have a look: https://github.com/llvm-hs/llvm-irbuilder/tree/Transformers

**Comment by cocreature** [comment]
Great work!

I’m not quite sure how I feel about removing the distinction between the block and the function level (I don’t think generating a lot of blocks is really a problem, LLVM is quite happy to remove those) but I don’t see any fundamental problem with removing it either so from my side removing it is fine.

**Comment by ollef** [comment]
Thanks!

It's more that I care about what the generated code looks like than efficiency since I sometimes have to read it to debug things. That's also why I like having a way to name things. :)

Next I'm planning to rewrite my backend using this so we'll see if there are any changes that need to be made for that.
